File: app/src/main/kotlin/com/foodshare/core/network/CircuitBreaker.kt
package com.foodshare.core.network

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong

/**
 * Circuit breaker states for fault tolerance.
 *
 * SYNC: This mirrors Swift FoodshareCore.CircuitState
 */
enum class CircuitState {
    /** Normal operation - requests flow through */
    CLOSED,
    /** Circuit tripped - requests fail fast */
    OPEN,
    /** Testing recovery - limited requests allowed */
    HALF_OPEN
}

/**
 * Circuit breaker pattern implementation for fault tolerance.
 *
 * Prevents cascading failures by failing fast when a service is unavailable.
 * State transitions: CLOSED → OPEN → HALF_OPEN → CLOSED
 *
 * SYNC: This mirrors Swift FoodshareCore.CircuitBreaker
 *
 * @param name Identifier for this circuit breaker (for logging)
 * @param failureThreshold Consecutive failures before opening circuit
 * @param resetTimeoutMs Time to wait before attempting recovery
 * @param halfOpenRequests Number of test requests in half-open state
 */
class CircuitBreaker(
    val name: String,
    private val failureThreshold: Int = 5,
    private val resetTimeoutMs: Long = 30_000,
    private val halfOpenRequests: Int = 3
) {
    private val mutex = Mutex()

    @Volatile
    private var state: CircuitState = CircuitState.CLOSED

    private val consecutiveFailures = AtomicInteger(0)
    private val halfOpenSuccesses = AtomicInteger(0)
    private val halfOpenAttempts = AtomicInteger(0)
    private val lastFailureTime = AtomicLong(0)

    // Metrics
    private val totalRequests = AtomicLong(0)
    private val totalFailures = AtomicLong(0)
    private val totalSuccesses = AtomicLong(0)

    /**
     * Current state of the circuit breaker.
     */
    val currentState: CircuitState
        get() = state

    /**
     * Circuit breaker metrics for monitoring.
     */
    data class Metrics(
        val state: CircuitState,
        val totalRequests: Long,
        val totalSuccesses: Long,
        val totalFailures: Long,
        val consecutiveFailures: Int,
        val failureRate: Double
    )

    /**
     * Get current metrics.
     */
    fun getMetrics(): Metrics {
        val total = totalRequests.get()
        val failures = totalFailures.get()
        return Metrics(
            state = state,
            totalRequests = total,
            totalSuccesses = totalSuccesses.get(),
            totalFailures = failures,
            consecutiveFailures = consecutiveFailures.get(),
            failureRate = if (total > 0) failures.toDouble() / total else 0.0
        )
    }

    /**
     * Check if a request can proceed.
     *
     * @return true if the request should be allowed, false if it should fail fast
     */
    suspend fun canProceed(): Boolean = mutex.withLock {
        totalRequests.incrementAndGet()

        when (state) {
            CircuitState.CLOSED -> true

            CircuitState.OPEN -> {
                val timeSinceFailure = System.currentTimeMillis() - lastFailureTime.get()
                if (timeSinceFailure >= resetTimeoutMs) {
                    // Transition to half-open
                    transitionTo(CircuitState.HALF_OPEN)
                    halfOpenAttempts.set(0)
                    halfOpenSuccesses.set(0)
                    true
                } else {
                    // Still in cooldown
                    false
                }
            }

            CircuitState.HALF_OPEN -> {
                // Allow limited requests in half-open state
                halfOpenAttempts.incrementAndGet() <= halfOpenRequests
            }
        }
    }

    /**
     * Record a successful request.
     */
    suspend fun recordSuccess() = mutex.withLock {
        totalSuccesses.incrementAndGet()

        when (state) {
            CircuitState.CLOSED -> {
                // Reset consecutive failures
                consecutiveFailures.set(0)
            }

            CircuitState.HALF_OPEN -> {
                val successes = halfOpenSuccesses.incrementAndGet()
                if (successes >= halfOpenRequests) {
                    // Recovery successful - close circuit
                    transitionTo(CircuitState.CLOSED)
                    consecutiveFailures.set(0)
                }
            }

            CircuitState.OPEN -> {
                // Shouldn't happen, but handle gracefully
            }
        }
    }

    /**
     * Record a failed request.
     */
    suspend fun recordFailure() = mutex.withLock {
        totalFailures.incrementAndGet()
        lastFailureTime.set(System.currentTimeMillis())

        when (state) {
            CircuitState.CLOSED -> {
                val failures = consecutiveFailures.incrementAndGet()
                if (failures >= failureThreshold) {
                    // Too many failures - open circuit
                    transitionTo(CircuitState.OPEN)
                }
            }

            CircuitState.HALF_OPEN -> {
                // Recovery failed - back to open
                transitionTo(CircuitState.OPEN)
            }

            CircuitState.OPEN -> {
                // Already open, update failure time
            }
        }
    }

    /**
     * Execute a request with circuit breaker protection.
     *
     * @param block The suspending function to execute
     * @return Result of the operation
     * @throws CircuitOpenException if circuit is open
     */
    suspend fun <T> execute(block: suspend () -> T): Result<T> {
        if (!canProceed()) {
            return Result.failure(CircuitOpenException(name, remainingCooldownMs()))
        }

        return try {
            val result = block()
            recordSuccess()
            Result.success(result)
        } catch (e: Exception) {
            recordFailure()
            Result.failure(e)
        }
    }

    /**
     * Force the circuit to a specific state (for testing/admin).
     */
    suspend fun forceState(newState: CircuitState) = mutex.withLock {
        transitionTo(newState)
        if (newState == CircuitState.CLOSED) {
            consecutiveFailures.set(0)
        }
    }

    /**
     * Reset the circuit breaker to initial state.
     */
    suspend fun reset() = mutex.withLock {
        state = CircuitState.CLOSED
        consecutiveFailures.set(0)
        halfOpenSuccesses.set(0)
        halfOpenAttempts.set(0)
        lastFailureTime.set(0)
    }

    /**
     * Get remaining cooldown time in milliseconds.
     */
    fun remainingCooldownMs(): Long {
        if (state != CircuitState.OPEN) return 0
        val elapsed = System.currentTimeMillis() - lastFailureTime.get()
        return maxOf(0, resetTimeoutMs - elapsed)
    }

    private fun transitionTo(newState: CircuitState) {
        val oldState = state
        state = newState
        onStateChange?.invoke(name, oldState, newState)
    }

    companion object {
        /**
         * Global state change listener for logging/monitoring.
         */
        var onStateChange: ((name: String, from: CircuitState, to: CircuitState) -> Unit)? = null
    }
}

/**
 * Exception thrown when circuit breaker is open.
 */
class CircuitOpenException(
    val circuitName: String,
    val remainingCooldownMs: Long
) : Exception("Circuit '$circuitName' is open. Retry in ${remainingCooldownMs}ms")

/**
 * Registry of circuit breakers for different operations.
 */
object CircuitBreakerRegistry {
    private val breakers = mutableMapOf<String, CircuitBreaker>()

    /**
     * Get or create a circuit breaker for the given name.
     */
    @Synchronized
    fun getOrCreate(
        name: String,
        failureThreshold: Int = 5,
        resetTimeoutMs: Long = 30_000,
        halfOpenRequests: Int = 3
    ): CircuitBreaker {
        return breakers.getOrPut(name) {
            CircuitBreaker(
                name = name,
                failureThreshold = failureThreshold,
                resetTimeoutMs = resetTimeoutMs,
                halfOpenRequests = halfOpenRequests
            )
        }
    }

    /**
     * Get a circuit breaker by name if it exists.
     */
    @Synchronized
    fun get(name: String): CircuitBreaker? = breakers[name]

    /**
     * Get all circuit breakers and their states.
     */
    @Synchronized
    fun getAllMetrics(): Map<String, CircuitBreaker.Metrics> {
        return breakers.mapValues { it.value.getMetrics() }
    }

    /**
     * Reset all circuit breakers.
     */
    suspend fun resetAll() {
        breakers.values.forEach { it.reset() }
    }

    /**
     * Clear all circuit breakers (for testing).
     */
    @Synchronized
    fun clear() {
        breakers.clear()
    }
}

File: Sources/FoodShare/Core/Networking/CircuitBreaker.swift
//
//  CircuitBreaker.swift
//  Foodshare
//
//  Enterprise circuit breaker pattern for resilient network operations
//
//  Prevents cascading failures by:
//  - Stopping requests to failing services
//  - Allowing recovery time for unhealthy services
//  - Gradually testing if services have recovered
//


#if !SKIP
import Foundation
import OSLog

// MARK: - Circuit Breaker

/// Enterprise-grade circuit breaker for network resilience
///
/// Implements the circuit breaker pattern with three states:
/// - **Closed**: Normal operation, requests flow through
/// - **Open**: Circuit tripped, requests fail fast
/// - **Half-Open**: Testing if service has recovered
///
/// Usage:
/// ```swift
/// let breaker = CircuitBreaker(name: "supabase-api")
///
/// do {
///     let result = try await breaker.execute {
///         try await apiClient.fetchProfile()
///     }
/// } catch CircuitBreakerError.circuitOpen {
///     // Service is known to be down, fail fast
///     showOfflineBanner()
/// }
/// ```
actor CircuitBreaker {
    // MARK: - Types

    /// Circuit breaker states
    enum State: String, Sendable {
        case closed // Normal operation
        case open // Failing fast
        case halfOpen // Testing recovery
    }

    /// Configuration for circuit breaker behavior
    struct Configuration: Sendable {
        /// Number of failures before opening the circuit
        let failureThreshold: Int

        /// Time to wait before attempting recovery (seconds)
        let resetTimeout: TimeInterval

        /// Number of successful calls needed to close from half-open
        let successThreshold: Int

        /// Window for counting failures (seconds)
        let failureWindow: TimeInterval

        /// Whether to track slow calls as failures
        let trackSlowCalls: Bool

        /// Threshold for slow call duration (seconds)
        let slowCallThreshold: TimeInterval

        /// Percentage of slow calls that triggers circuit open
        let slowCallRateThreshold: Double

        static let `default` = Configuration(
            failureThreshold: 5,
            resetTimeout: 30,
            successThreshold: 3,
            failureWindow: 60,
            trackSlowCalls: true,
            slowCallThreshold: 5.0,
            slowCallRateThreshold: 0.5,
        )

        static let aggressive = Configuration(
            failureThreshold: 3,
            resetTimeout: 60,
            successThreshold: 5,
            failureWindow: 30,
            trackSlowCalls: true,
            slowCallThreshold: 3.0,
            slowCallRateThreshold: 0.3,
        )

        static let lenient = Configuration(
            failureThreshold: 10,
            resetTimeout: 15,
            successThreshold: 2,
            failureWindow: 120,
            trackSlowCalls: false,
            slowCallThreshold: 10.0,
            slowCallRateThreshold: 0.7,
        )
    }

    /// Statistics about circuit breaker performance
    struct Statistics: Sendable {
        let name: String
        let state: State
        let totalCalls: Int
        let successfulCalls: Int
        let failedCalls: Int
        let rejectedCalls: Int
        let slowCalls: Int
        let averageResponseTime: TimeInterval
        let lastFailure: Date?
        let lastSuccess: Date?
        let stateChangedAt: Date

        var successRate: Double {
            guard totalCalls > 0 else { return 1.0 }
            return Double(successfulCalls) / Double(totalCalls)
        }

        var failureRate: Double {
            guard totalCalls > 0 else { return 0.0 }
            return Double(failedCalls) / Double(totalCalls)
        }
    }

    // MARK: - Properties

    let name: String
    private let config: Configuration
    private let logger: Logger

    private var state: State = .closed
    private var stateChangedAt = Date()

    // Failure tracking
    private var failureCount = 0
    private var failureTimestamps: [Date] = []
    private var lastFailure: Date?

    // Success tracking (for half-open state)
    private var halfOpenSuccessCount = 0
    private var lastSuccess: Date?

    // Statistics
    private var totalCalls = 0
    private var successfulCalls = 0
    private var failedCalls = 0
    private var rejectedCalls = 0
    private var slowCalls = 0
    private var responseTimes: [TimeInterval] = []

    // MARK: - Initialization

    init(name: String, config: Configuration = .default) {
        self.name = name
        self.config = config
        self.logger = Logger(subsystem: Constants.bundleIdentifier, category: "CircuitBreaker[\(name)]")
    }

    // MARK: - Execute

    /// Execute an operation through the circuit breaker
    ///
    /// - Parameter operation: The async operation to execute
    /// - Returns: The result of the operation
    /// - Throws: `CircuitBreakerError.circuitOpen` if circuit is open,
    ///           or the original error if the operation fails
    func execute<T>(_ operation: () async throws -> T) async throws -> T {
        totalCalls += 1

        // Check if we should allow the request
        switch state {
        case .open:
            if shouldAttemptReset() {
                logger.info("Transitioning to half-open state")
                transitionTo(.halfOpen)
            } else {
                rejectedCalls += 1
                logger.debug("Circuit open - rejecting request")
                throw CircuitBreakerError.circuitOpen(name: name, retryAfter: timeUntilReset())
            }

        case .halfOpen:
            // Allow limited requests through for testing
            break

        case .closed:
            // Normal operation
            break
        }

        // Execute the operation
        let startTime = Date()
        do {
            let result = try await operation()
            recordSuccess(duration: Date().timeIntervalSince(startTime))
            return result
        } catch {
            recordFailure(error: error, duration: Date().timeIntervalSince(startTime))
            throw error
        }
    }

    // MARK: - State Management

    private func recordSuccess(duration: TimeInterval) {
        successfulCalls += 1
        lastSuccess = Date()
        responseTimes.append(duration)

        // Trim response times array
        if responseTimes.count > 100 {
            responseTimes.removeFirst(responseTimes.count - 100)
        }

        // Check for slow call
        if config.trackSlowCalls, duration > config.slowCallThreshold {
            slowCalls += 1
            logger.warning("Slow call detected: \(String(format: "%.2f", duration))s")
        }

        switch state {
        case .halfOpen:
            halfOpenSuccessCount += 1
            if halfOpenSuccessCount >= config.successThreshold {
                logger.info("Circuit recovered - closing")
                transitionTo(.closed)
            }

        case .closed:
            // Reset failure count on success in closed state
            cleanupOldFailures()

        case .open:
            // Shouldn't happen, but handle gracefully
            break
        }
    }

    private func recordFailure(error: Error, duration: TimeInterval) {
        failedCalls += 1
        lastFailure = Date()
        responseTimes.append(duration)

        switch state {
        case .closed:
            failureCount += 1
            failureTimestamps.append(Date())
            cleanupOldFailures()

            logger
                .warning(
                    "Failure recorded (\(self.failureCount)/\(self.config.failureThreshold)): \(error.localizedDescription)",
                )

            if failureCount >= config.failureThreshold {
                logger.error("Failure threshold reached - opening circuit")
                transitionTo(.open)
            }

        case .halfOpen:
            logger.warning("Failure in half-open state - reopening circuit")
            transitionTo(.open)

        case .open:
            // Already open
            break
        }
    }

    private func transitionTo(_ newState: State) {
        let oldState = state
        state = newState
        stateChangedAt = Date()

        // Reset counters on state change
        switch newState {
        case .closed:
            failureCount = 0
            failureTimestamps.removeAll()
            halfOpenSuccessCount = 0

        case .open:
            halfOpenSuccessCount = 0

        case .halfOpen:
            halfOpenSuccessCount = 0
        }

        logger.info("State changed: \(oldState.rawValue) -> \(newState.rawValue)")
    }

    private func shouldAttemptReset() -> Bool {
        guard state == .open else { return false }
        return Date().timeIntervalSince(stateChangedAt) >= config.resetTimeout
    }

    private func timeUntilReset() -> TimeInterval {
        guard state == .open else { return 0 }
        let elapsed = Date().timeIntervalSince(stateChangedAt)
        return max(0.0, config.resetTimeout - elapsed)
    }

    private func cleanupOldFailures() {
        let cutoff = Date().addingTimeInterval(-config.failureWindow)
        failureTimestamps.removeAll { $0 < cutoff }
        failureCount = failureTimestamps.count
    }

    // MARK: - Manual Controls

    /// Force the circuit to open (for testing or manual intervention)
    func forceOpen() {
        logger.warning("Circuit manually opened")
        transitionTo(.open)
    }

    /// Force the circuit to close (for testing or manual intervention)
    func forceClose() {
        logger.info("Circuit manually closed")
        transitionTo(.closed)
    }

    /// Reset all statistics and state
    func reset() {
        transitionTo(.closed)
        totalCalls = 0
        successfulCalls = 0
        failedCalls = 0
        rejectedCalls = 0
        slowCalls = 0
        responseTimes.removeAll()
        lastFailure = nil
        lastSuccess = nil
        logger.info("Circuit breaker reset")
    }

    // MARK: - Statistics

    /// Get current statistics
    var statistics: Statistics {
        Statistics(
            name: name,
            state: state,
            totalCalls: totalCalls,
            successfulCalls: successfulCalls,
            failedCalls: failedCalls,
            rejectedCalls: rejectedCalls,
            slowCalls: slowCalls,
            averageResponseTime: responseTimes.isEmpty ? 0 : responseTimes.reduce(0, +) / Double(responseTimes.count),
            lastFailure: lastFailure,
            lastSuccess: lastSuccess,
            stateChangedAt: stateChangedAt,
        )
    }

    /// Current state of the circuit
    var currentState: State {
        state
    }

    /// Whether requests are currently being allowed
    var isAllowingRequests: Bool {
        switch state {
        case .closed, .halfOpen:
            true
        case .open:
            shouldAttemptReset()
        }
    }

    // MARK: - Public Outcome Recording

    /// Record a successful operation outcome (for use when not using execute())
    /// - Parameter duration: The duration of the operation
    func reportSuccess(duration: TimeInterval) {
        recordSuccess(duration: duration)
    }

    /// Record a failed operation outcome (for use when not using execute())
    /// - Parameters:
    ///   - error: The error that occurred
    ///   - duration: The duration of the operation
    func reportFailure(error: Error, duration: TimeInterval) {
        recordFailure(error: error, duration: duration)
    }

    /// Prepares to execute a request - returns whether request should proceed
    /// Call this before executing an operation, then call reportSuccess/reportFailure after
    /// - Returns: true if the request can proceed, false if circuit is open
    func prepareRequest() -> Bool {
        totalCalls += 1

        switch state {
        case .open:
            if shouldAttemptReset() {
                logger.info("Transitioning to half-open state")
                transitionTo(.halfOpen)
                return true
            } else {
                rejectedCalls += 1
                logger.debug("Circuit open - rejecting request")
                return false
            }

        case .halfOpen, .closed:
            return true
        }
    }
}

// MARK: - Circuit Breaker Error

/// Errors thrown by the circuit breaker
enum CircuitBreakerError: Error, LocalizedError {
    case circuitOpen(name: String, retryAfter: TimeInterval)

    var errorDescription: String? {
        switch self {
        case let .circuitOpen(name, retryAfter):
            "Service '\(name)' is temporarily unavailable. Retry in \(Int(retryAfter)) seconds."
        }
    }

    var retryAfter: TimeInterval? {
        switch self {
        case let .circuitOpen(_, retryAfter):
            retryAfter
        }
    }
}

// MARK: - Circuit Breaker Registry

/// Global registry for managing multiple circuit breakers
actor CircuitBreakerRegistry {
    static let shared = CircuitBreakerRegistry()

    private var breakers: [String: CircuitBreaker] = [:]

    private init() {}

    /// Get or create a circuit breaker for a service
    func breaker(for service: String, config: CircuitBreaker.Configuration = .default) -> CircuitBreaker {
        if let existing = breakers[service] {
            return existing
        }

        let newBreaker = CircuitBreaker(name: service, config: config)
        breakers[service] = newBreaker
        return newBreaker
    }

    /// Get all registered circuit breakers
    var allBreakers: [CircuitBreaker] {
        Array(breakers.values)
    }

    /// Get statistics for all circuit breakers
    func allStatistics() async -> [CircuitBreaker.Statistics] {
        var stats: [CircuitBreaker.Statistics] = []
        for breaker in breakers.values {
            await stats.append(breaker.statistics)
        }
        return stats
    }

    /// Reset all circuit breakers
    func resetAll() async {
        for breaker in breakers.values {
            await breaker.reset()
        }
    }
}

// MARK: - Convenience Extension

extension CircuitBreaker {
    /// Pre-configured circuit breaker for Supabase API calls
    static let supabase = CircuitBreaker(name: "supabase-api", config: .default)

    /// Pre-configured circuit breaker for image loading
    static let images = CircuitBreaker(name: "images", config: .lenient)

    /// Pre-configured circuit breaker for real-time connections
    static let realtime = CircuitBreaker(name: "realtime", config: .aggressive)
}

#endif
